File: Backend/BudgItUp/src/main/java/org/g3_inge4_isi_en/budgitup/service/CategoryService.java
// Backend/Budgitup/src/main/java/org/g3_inge4_isi_en/budgitup/service/CategoryService.java
// COPY AND PASTE THIS ENTIRE FILE - REPLACES YOUR EXISTING CategoryService.java

package org.g3_inge4_isi_en.budgitup.service;

import org.g3_inge4_isi_en.budgitup.dto.CategoryDto;
import org.g3_inge4_isi_en.budgitup.entity.Category;
import org.g3_inge4_isi_en.budgitup.entity.User;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.g3_inge4_isi_en.budgitup.repository.CategoryRepository;
import org.g3_inge4_isi_en.budgitup.repository.UserRepository;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CategoryService {

    private final CategoryRepository categoryRepository;
    private final UserRepository userRepository;

    /**
     * List system default categories (for signup selection only)
     * These have isDefault=true and user_id=null
     */
    public List<CategoryDto> listDefaultCategories() {
        return categoryRepository.findByIsDefaultTrue().stream()
                .filter(c -> c.getUser() == null) // Only system defaults
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    /**
     * List categories for a specific user only
     * Returns ONLY categories that belong to this user
     * Includes both:
     *  - Chosen defaults (isDefault=true, user_id=userId) - can't delete
     *  - Truly custom (isDefault=false, user_id=userId) - can delete
     */
    public List<CategoryDto> listUserCategories(Long userId) {
        return categoryRepository.findByUserId(userId).stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    /**
     * Create a truly custom category for a user
     * These have isDefault=false and can be deleted by user
     */
    @Transactional
    public CategoryDto createCustomCategory(Long userId, CategoryDto dto) {
        if (dto.getName() == null || dto.getName().trim().isEmpty()) {
            throw new IllegalArgumentException("Category name required");
        }

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new EntityNotFoundException("User not found"));

        // Check uniqueness: only within this user's categories
        List<Category> userCats = categoryRepository.findByUserId(userId);
        boolean exists = userCats.stream()
                .anyMatch(c -> c.getName().equalsIgnoreCase(dto.getName().trim()));

        if (exists) {
            throw new IllegalArgumentException("You already have a category with that name");
        }

        Category cat = Category.builder()
                .name(dto.getName().trim())
                .color(dto.getColor())
                .icon(dto.geticon())
                .isDefault(false) // Truly custom categories are NOT default (can be deleted)
                .user(user)
                .build();

        Category saved = categoryRepository.save(cat);
        return toDto(saved);
    }

    /**
     * Update a category
     * Ensures the category belongs to this user
     * Can update both chosen defaults and truly custom categories
     */
    @Transactional
    public CategoryDto updateCustomCategory(Long userId, Long categoryId, CategoryDto dto) {
        Category cat = categoryRepository.findByIdAndUserId(categoryId, userId)
                .orElseThrow(() -> new EntityNotFoundException("Category not found or does not belong to you"));

        // If renaming, check uniqueness within user's categories (exclude itself)
        if (!cat.getName().equalsIgnoreCase(dto.getName())) {
            List<Category> userCats = categoryRepository.findByUserId(userId);
            boolean exists = userCats.stream()
                    .filter(c -> !c.getId().equals(categoryId))
                    .anyMatch(c -> c.getName().equalsIgnoreCase(dto.getName()));

            if (exists) {
                throw new IllegalArgumentException("You already have a category with that name");
            }
            cat.setName(dto.getName().trim());
        }

        cat.setColor(dto.getColor());
        if (dto.geticon() != null) {
            cat.setIcon(dto.geticon());
        }

        Category updated = categoryRepository.save(cat);
        return toDto(updated);
    }

    /**
     * Delete a category
     * ONLY allows deletion of truly custom categories (isDefault=false)
     * Prevents deletion of user's chosen defaults (isDefault=true)
     */
    @Transactional
    public void deleteCustomCategory(Long userId, Long categoryId) {
        Category cat = categoryRepository.findByIdAndUserId(categoryId, userId)
                .orElseThrow(() -> new EntityNotFoundException("Category not found or does not belong to you"));

        // IMPORTANT: Prevent deletion of chosen defaults
        if (cat.isDefault()) {
            throw new IllegalArgumentException("Cannot delete default categories. You can only delete custom categories you've created.");
        }

        // Cascade delete is handled by @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
        categoryRepository.delete(cat);
    }

    private CategoryDto toDto(Category c) {
        CategoryDto dto = new CategoryDto();
        dto.setId(c.getId());
        dto.setName(c.getName());
        dto.setColor(c.getColor());
        dto.seticon(c.getIcon());
        dto.setDefault(c.isDefault()); // TRUE = chosen default, FALSE = truly custom
        return dto;
    }
}

File: Frontend/BudgItUp/services/database/CategoryService.ts


import { getDatabase } from './schema';

export interface CategoryDto {
    id?: number;
    name: string;
    color: string;
    icon: string;
    isDefault: boolean;
}

class CategoryService {
    // Get default categories (system templates)
    async getDefaultCategories(): Promise<CategoryDto[]> {
        try {
            const db = await getDatabase(); // This now auto-initializes

            const rows = await db.getAllAsync<{
                id: number;
                name: string;
                color: string;
                icon: string;
                is_default: number;
            }>(
                'SELECT id, name, color, icon, is_default FROM categories WHERE is_default = 1 AND user_id IS NULL ORDER BY name'
            );

            return rows.map(row => ({
                id: row.id,
                name: row.name,
                color: row.color,
                icon: row.icon,
                isDefault: row.is_default === 1
            }));
        } catch (error: any) {
            console.error('❌ Get default categories error:', error);
            throw error;
        }
    }

    // Get user's categories (both chosen defaults and custom)
    async getUserCategories(userId: number): Promise<CategoryDto[]> {
        try {
            const db = await getDatabase();

            const rows = await db.getAllAsync<{
                id: number;
                name: string;
                color: string;
                icon: string;
                is_default: number;
            }>(
                'SELECT id, name, color, icon, is_default FROM categories WHERE user_id = ? ORDER BY name',
                [userId]
            );

            return rows.map(row => ({
                id: row.id,
                name: row.name,
                color: row.color,
                icon: row.icon,
                isDefault: row.is_default === 1
            }));
        } catch (error: any) {
            console.error('❌ Get user categories error:', error);
            throw error;
        }
    }

    // Create custom category
    async createCategory(userId: number, data: Partial<CategoryDto>): Promise<CategoryDto> {
        try {
            const db = await getDatabase();

            // Check for duplicate name within user's categories
            const existing = await db.getFirstAsync<{ id: number }>(
                'SELECT id FROM categories WHERE user_id = ? AND LOWER(name) = LOWER(?)',
                [userId, data.name || '']
            );

            if (existing) {
                throw new Error('You already have a category with that name');
            }

            // Insert new category
            const result = await db.runAsync(
                'INSERT INTO categories (name, color, icon, is_default, user_id) VALUES (?, ?, ?, 0, ?)',
                [data.name || '', data.color || '#348DDB', data.icon || 'pricetag-outline', userId]
            );

            return {
                id: result.lastInsertRowId,
                name: data.name || '',
                color: data.color || '#348DDB',
                icon: data.icon || 'pricetag-outline',
                isDefault: false
            };
        } catch (error: any) {
            console.error('❌ Create category error:', error);
            throw error;
        }
    }

    // Update category
    async updateCategory(
        userId: number,
        categoryId: number,
        data: Partial<CategoryDto>
    ): Promise<CategoryDto> {
        try {
            const db = await getDatabase();

            // Verify category belongs to user
            const category = await db.getFirstAsync<{
                id: number;
                name: string;
                is_default: number;
            }>(
                'SELECT id, name, is_default FROM categories WHERE id = ? AND user_id = ?',
                [categoryId, userId]
            );

            if (!category) {
                throw new Error('Category not found or does not belong to user');
            }

            // Check for duplicate name (excluding current category)
            if (data.name && data.name !== category.name) {
                const duplicate = await db.getFirstAsync<{ id: number }>(
                    'SELECT id FROM categories WHERE user_id = ? AND LOWER(name) = LOWER(?) AND id != ?',
                    [userId, data.name, categoryId]
                );

                if (duplicate) {
                    throw new Error('You already have a category with that name');
                }
            }

            // Update category
            const setClauses: string[] = [];
            const values: any[] = [];

            if (data.name !== undefined) {
                setClauses.push('name = ?');
                values.push(data.name);
            }
            if (data.color !== undefined) {
                setClauses.push('color = ?');
                values.push(data.color);
            }
            if (data.icon !== undefined) {
                setClauses.push('icon = ?');
                values.push(data.icon);
            }

            if (setClauses.length > 0) {
                values.push(categoryId, userId);
                await db.runAsync(
                    `UPDATE categories SET ${setClauses.join(', ')} WHERE id = ? AND user_id = ?`,
                    values
                );
            }

            // Fetch updated category
            const updated = await db.getFirstAsync<{
                id: number;
                name: string;
                color: string;
                icon: string;
                is_default: number;
            }>(
                'SELECT id, name, color, icon, is_default FROM categories WHERE id = ?',
                [categoryId]
            );

            if (!updated) {
                throw new Error('Category not found after update');
            }

            return {
                id: updated.id,
                name: updated.name,
                color: updated.color,
                icon: updated.icon,
                isDefault: updated.is_default === 1
            };
        } catch (error: any) {
            console.error('❌ Update category error:', error);
            throw error;
        }
    }

    // Delete category (only non-default)
    async deleteCategory(userId: number, categoryId: number): Promise<void> {
        try {
            const db = await getDatabase();

            // Verify category belongs to user and is not default
            const category = await db.getFirstAsync<{
                id: number;
                is_default: number;
            }>(
                'SELECT id, is_default FROM categories WHERE id = ? AND user_id = ?',
                [categoryId, userId]
            );

            if (!category) {
                throw new Error('Category not found or does not belong to user');
            }

            if (category.is_default === 1) {
                throw new Error('Cannot delete default categories');
            }

            // Delete category (cascades to expenses, incomes, budgets)
            await db.runAsync(
                'DELETE FROM categories WHERE id = ? AND user_id = ?',
                [categoryId, userId]
            );

            console.log('✅ Category deleted:', categoryId);
        } catch (error: any) {
            console.error('❌ Delete category error:', error);
            throw error;
        }
    }
}

export default new CategoryService();